File: APPLICATION/models/venta.py
"""
venta.py — Modelo de Venta con serialización binaria de longitud fija.

Estructura del registro (52 bytes por registro):
┌──────────┬──────────────┬──────────────┬──────────────┬──────────────┬──────────────┬──────────┐
│  id      │  cliente_id  │  fecha       │  subtotal    │  igv         │  total       │  estado  │
│  4 bytes │  4 bytes     │  19 bytes    │  8 bytes     │  8 bytes     │  8 bytes     │  1 byte  │
└──────────┴──────────────┴──────────────┴──────────────┴──────────────┴──────────────┴──────────┘
Total: 4 + 4 + 19 + 8 + 8 + 8 + 1 = 52 bytes

Formato de fecha: "YYYY-MM-DD HH:MM:SS" (19 caracteres ASCII)
"""
import struct
from dataclasses import dataclass, field
from datetime import datetime
from typing import ClassVar


@dataclass
class Venta:
    """
    Representa una transacción de venta realizada en el POS.

    Serialización: struct big-endian "!II19sdddB"
    Tamaño fijo:   52 bytes por registro
    """

    # ! = big-endian
    # I   = unsigned int (4 bytes)  × 2 → id, cliente_id
    # 19s = char[19]    (19 bytes)      → fecha "YYYY-MM-DD HH:MM:SS"
    # d   = double      (8 bytes)  × 3 → subtotal, igv, total
    # B   = unsigned char (1 byte)      → estado
    FORMAT: ClassVar[str] = "!II19sdddB"
    SIZE:   ClassVar[int] = struct.calcsize("!II19sdddB")  # 52

    id:         int   = 0
    cliente_id: int   = 0
    fecha:      str   = field(
        default_factory=lambda: datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    )
    subtotal:   float = 0.0
    igv:        float = 0.0
    total:      float = 0.0
    estado:     int   = 1   # 1=activo, 0=anulado

    def pack(self) -> bytes:
        """Serializa la venta a 52 bytes de longitud fija."""
        fecha_b = self.fecha.encode("utf-8")[:19].ljust(19, b"\x00")
        return struct.pack(
            self.FORMAT,
            self.id,
            self.cliente_id,
            fecha_b,
            self.subtotal,
            self.igv,
            self.total,
            self.estado,
        )

    @classmethod
    def unpack(cls, data: bytes) -> "Venta":
        """Deserializa 52 bytes a un objeto Venta."""
        if len(data) != cls.SIZE:
            raise ValueError(
                f"Venta.unpack: se esperaban {cls.SIZE} bytes, "
                f"se recibieron {len(data)}"
            )
        raw = struct.unpack(cls.FORMAT, data)
        return cls(
            id=raw[0],
            cliente_id=raw[1],
            fecha=raw[2].rstrip(b"\x00").decode("utf-8", errors="replace"),
            subtotal=raw[3],
            igv=raw[4],
            total=raw[5],
            estado=raw[6],
        )

    def to_dict(self) -> dict:
        return {
            "id": self.id, "cliente_id": self.cliente_id, "fecha": self.fecha,
            "subtotal": self.subtotal, "igv": self.igv,
            "total": self.total, "estado": self.estado,
        }

    def to_csv_row(self) -> str:
        return (
            f"{self.id},{self.cliente_id},{self.fecha},"
            f"{self.subtotal:.2f},{self.igv:.2f},{self.total:.2f},{self.estado}"
        )

    def __repr__(self) -> str:
        return (
            f"Venta(id={self.id}, cliente_id={self.cliente_id}, "
            f"fecha={self.fecha!r}, total=S/{self.total:.2f}, estado={self.estado})"
        )

File: APPLICATION/src/main/java/com/minimarket/models/Venta.java
package com.minimarket.models;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Modelo de Venta con serialización binaria de longitud fija.
 *
 * Estructura del registro (52 bytes):
 *   id(4B) + cliente_id(4B) + fecha(19B) + subtotal(8B) + igv(8B) + total(8B) + estado(1B)
 *
 * Compatible con Python: struct.pack("!II19sdddB", ...)
 * Formato de fecha: "YYYY-MM-DD HH:MM:SS" (19 caracteres ASCII)
 */
public class Venta implements BinaryRecord {

    public static final int SIZE = 52; // 4 + 4 + 19 + 8 + 8 + 8 + 1

    static final DateTimeFormatter FECHA_FMT =
        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    private int    id;
    private int    clienteId;
    private String fecha;
    private double subtotal;
    private double igv;
    private double total;
    private int    estado;

    public Venta() {
        this.fecha  = LocalDateTime.now().format(FECHA_FMT);
        this.estado = 1;
    }

    public Venta(int id, int clienteId, String fecha,
                 double subtotal, double igv, double total, int estado) {
        this.id        = id;
        this.clienteId = clienteId;
        this.fecha     = fecha != null ? fecha : LocalDateTime.now().format(FECHA_FMT);
        this.subtotal  = subtotal;
        this.igv       = igv;
        this.total     = total;
        this.estado    = estado;
    }

    // ── Serialización ────────────────────────────────────────────────────────

    @Override
    public byte[] toBytes() {
        ByteArrayOutputStream bos = new ByteArrayOutputStream(SIZE);
        DataOutputStream dos = new DataOutputStream(bos);
        try {
            dos.writeInt(id);
            dos.writeInt(clienteId);
            writeFixed(dos, fecha, 19);
            dos.writeDouble(subtotal);
            dos.writeDouble(igv);
            dos.writeDouble(total);
            dos.writeByte(estado);
        } catch (IOException e) {
            throw new RuntimeException("Error serializando Venta", e);
        }
        return bos.toByteArray();
    }

    public static Venta fromBytes(byte[] data) throws IOException {
        if (data.length != SIZE) {
            throw new IllegalArgumentException(
                "Venta.fromBytes: se esperaban " + SIZE + " bytes, se recibieron " + data.length);
        }
        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(data));
        Venta v = new Venta();
        v.id        = dis.readInt();
        v.clienteId = dis.readInt();
        v.fecha     = readFixed(dis, 19);
        v.subtotal  = dis.readDouble();
        v.igv       = dis.readDouble();
        v.total     = dis.readDouble();
        v.estado    = dis.readByte() & 0xFF;
        return v;
    }

    // ── Exportación ──────────────────────────────────────────────────────────

    public String toCsvRow() {
        return String.format("%d,%d,%s,%.2f,%.2f,%.2f,%d",
            id, clienteId, fecha, subtotal, igv, total, estado);
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private static void writeFixed(DataOutputStream dos, String s, int len) throws IOException {
        byte[] src = (s != null ? s : "").getBytes(StandardCharsets.UTF_8);
        byte[] dst = new byte[len];
        System.arraycopy(src, 0, dst, 0, Math.min(src.length, len));
        dos.write(dst);
    }

    private static String readFixed(DataInputStream dis, int len) throws IOException {
        byte[] bytes = new byte[len];
        dis.readFully(bytes);
        int end = 0;
        while (end < len && bytes[end] != 0) end++;
        return new String(bytes, 0, end, StandardCharsets.UTF_8).strip();
    }

    // ── BinaryRecord ─────────────────────────────────────────────────────────

    @Override public int getId()           { return id; }
    @Override public void setId(int id)    { this.id = id; }
    @Override public int getEstado()       { return estado; }
    @Override public void setEstado(int e) { this.estado = e; }

    // ── Getters / Setters ────────────────────────────────────────────────────

    public int    getClienteId()            { return clienteId; }
    public void   setClienteId(int v)       { clienteId = v; }
    public String getFecha()                { return fecha; }
    public void   setFecha(String v)        { fecha = v; }
    public double getSubtotal()             { return subtotal; }
    public void   setSubtotal(double v)     { subtotal = v; }
    public double getIgv()                  { return igv; }
    public void   setIgv(double v)          { igv = v; }
    public double getTotal()                { return total; }
    public void   setTotal(double v)        { total = v; }

    @Override
    public String toString() {
        return String.format("Venta(id=%d, clienteId=%d, fecha='%s', total=S/%.2f, estado=%d)",
            id, clienteId, fecha, total, estado);
    }
}
